File: day16_Dragon_Checksum/python/part2.py
def shuffle(a):
    b = []
    for i in reversed(a):
        b.append((abs(int(i)-1)))
    return a+[0]+b

def checksum(a):
    b = []
    while len(b)%2 == 0:
        b = []
        for i,j in zip(a[::2], a[1::2]):
            if i == j:
                b.append(1)
            else:
                b.append(0)
        a=b
    return b

def run(puzzle_input = "11110010111001001", length=35651584):
    input_list = [int(i) for i in puzzle_input]
    while len(input_list)< length:
        input_list = shuffle(input_list)
    return checksum(input_list[:length])


if __name__ == "__main__":
    print("".join(str(i) for i in run()))

File: day16_Dragon_Checksum/cpp/part2.cpp
#include <iostream>
#include <vector>
#include <chrono>

using namespace std;
using namespace std::chrono;

vector<int> shuffle(vector<int> &input_vector){
    vector<int> output_vector (input_vector.size()*2+1, 0);
    int i = 0;
    int j = output_vector.size()-1;
    for(;i<input_vector.size();){
        output_vector[i] = input_vector[i];
        output_vector[j] = abs(input_vector[i]-1);
        i++;
        j--;
    }
    return output_vector;
}

vector<int> checksum(vector<int> &data){
    vector<int> rtn;
    while (data.size()%2 == 0){
        for (int i = 0; i < data.size(); i+=2){
            if (data[i]== data[i+1]){
                rtn.push_back(1);
            } else {
                rtn.push_back(0);
            }
        }
        data = rtn;
        rtn.clear();
    }
    return data;
}

void run(){
    int length_needed = 35651584;
    char input[] = {"11110010111001001"};
    vector<int> input_vector;
    for (int i =0; i < sizeof(input)-1;++i){
        input_vector.push_back((int)input[i] - 48);
    }
    while (input_vector.size() < length_needed){
        input_vector = shuffle(input_vector);
    }
    input_vector.resize(length_needed);
    vector<int> check_sum = checksum(input_vector);
    for (auto i:check_sum){
        cout << i;
    }
    cout << endl;
}

int main(){
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    run();
    high_resolution_clock::time_point t2 = high_resolution_clock::now();

    auto duration = duration_cast<milliseconds>( t2 - t1 ).count();

    cout << duration << endl;
}
